File: 019/cpp/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *first = head;
        for (int i = 0; i < n; i++, first = first->next) ;

        if (first == nullptr) {
            return head->next;
        }
        first = first->next;

        ListNode *second = head;
        for (; first != nullptr; first = first->next, second = second->next) ;

        second->next = second->next->next;
        return head;
    }
};

File: 019/rust/main.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut head = head;
        let mut first = head.as_deref();
        for i in 0..n {
            first = first.unwrap().next.as_deref();
        }

        if first.is_none() {
            return head.unwrap().next;
        }
        first = first.unwrap().next.as_deref();

        let mut i = 0;
        while first.is_some() {
            first = first.unwrap().next.as_deref();
            i += 1;
        }

        let mut second = head.as_deref_mut();
        while i > 0 {
            second = second.unwrap().next.as_deref_mut();
            i -= 1;
        }

        let second = second.unwrap();
        second.next = second.clone().next.unwrap().next;

        // match second {
        //     Some(ref mut second) => second.next = second.clone().next.unwrap().next,
        //     _ => (),
        // }
        return head;
    }
}
